File: mezzanine-compiler/src/main/java/com/anthonycr/mezzanine/filter/SupportedElementFilter.java
package com.anthonycr.mezzanine.filter;

import com.anthonycr.mezzanine.utils.MessagerUtils;

import org.jetbrains.annotations.NotNull;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;

import io.reactivex.functions.Predicate;

/**
 * Created by anthonycr on 5/22/17.
 */
public final class SupportedElementFilter {

    private SupportedElementFilter() {}

    @NotNull
    public static Predicate<Element> filterForSupportedElements() {
        return methodElement -> {

            Element classElement = methodElement.getEnclosingElement();

            if (classElement.getKind() != ElementKind.INTERFACE) {
                MessagerUtils.reportError(classElement, "Only interfaces are supported");
                return false;
            }

            if (classElement.getEnclosedElements().size() != 1) {
                MessagerUtils.reportError(classElement, "Only interfaces with 1 method are supported");
                return false;
            }

            if (!(methodElement instanceof ExecutableElement)) {
                MessagerUtils.reportError(methodElement, "Only method annotations are supported");
                return false;
            }

            ExecutableElement creatorMethodElement = (ExecutableElement) methodElement;

            if (!creatorMethodElement.getReturnType().toString().equals(String.class.getName().replace('$', '.'))) {
                MessagerUtils.reportError(creatorMethodElement, "Interface must have a String return type");
                return false;
            }

            return true;
        };
    }
}

File: mezzanine-compiler/src/main/java/com/anthonycr/mezzanine/filter/SupportedElementFilter.kt
package com.anthonycr.mezzanine.filter

import com.anthonycr.mezzanine.utils.MessagerUtils
import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.Modifier
import com.squareup.kotlinpoet.asTypeName
import com.squareup.kotlinpoet.ksp.toTypeName

/**
 * Filters a list of elements for only the elements supported by Mezzanine. If unsupported elements
 * are in the stream, it will report a message to the [MessagerUtils] so that the consumer knows
 * what they have done wrong.
 */
object SupportedElementFilter : (KSAnnotated) -> Boolean {

    override fun invoke(p1: KSAnnotated): Boolean {

        if (p1 !is KSClassDeclaration) {
            MessagerUtils.reportError(p1, "Only interfaces are supported")
            return false
        }


        if (p1.getDeclaredFunctions().toList().size != 1) {
             MessagerUtils.reportError(p1, "Only interfaces with 1 method are supported")
            return false
        }

        val firstFunction = p1.getDeclaredFunctions().first()

        if (firstFunction.parameters.isNotEmpty()) {
            MessagerUtils.reportError(p1, "Functions cannot have parameters")
            return false
        }

        if (firstFunction.returnType?.toTypeName() != String::class.asTypeName()) {
            MessagerUtils.reportError(p1, "Function must have return type string")
            return false
        }

        return true
    }

}
